libraries/Microsoft.Bot.Connector/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Connector.Authentication
+{
+    public class ThrottleException : Exception
+    {
+        public RetryParams RetryParams { get; set; }
+    }
+    
+    public class AdalAuthenticator 
+    {
+        // Our ADAL context. Acquires tokens and manages token caching for us.
+        private readonly AuthenticationContext authContext;
+
+        // Semaphore to control concurrency while refreshing tokens from ADAL.
+        // Whenever a token expires, we want only one request to retrieve a token. 
+        // Cached requests take less than 0.1 millisecond to resolve, so the semaphore doesn't hurt performance under load tests
+        // unless we have more than 10,000 requests per second, but in that case other things would break first.
+        private static Semaphore tokenRefreshSemaphore = new Semaphore(1, 1);
+        private static readonly TimeSpan SemaphoreTimeout = TimeSpan.FromSeconds(10);
```

Comment:
Rather than a `Semaphore`, might I suggest a `ReaderWriterLockSlim`? It should simplify the code, allow more concurrency for reads and improve overall performance.

Reply:
Good point. Initially the idea was to limit the number of concurrent threads within the critical section to a number greater than 1, which is why a semaphore was introduced. I'll try it out and run load tests on it and make sure that we are good performance wise.

Reply:
This is kind of moot now given other conversations going on so I'm going to resolve it.